File: internal/api/export.go
package api

import (
	"fmt"
	"log/slog"
	"mime"
	"net/http"

	"github.com/geoah/substrate/internal/substrate"
)

// exportRoute is the owner's recovery export, a non-record endpoint at the
// version root (decision 0066).
const exportRoute = "/export"

// getExport is GET /api/v1/export: the repository's directory as of one
// committed point, streamed as a tar (`application/x-tar`) laid out as a data
// root, with `snapshot.json` recording the point as its last entry. The
// bearer token is the whole credential: a token already reads every record
// and blob the archive carries, and its sealed files are ciphertext under a
// key the archive does not hold.
//
// The point is pinned before the status is written, so a refusal (a read-only
// process, a directory latched behind the tables) is an ordinary error body.
// Once the point is pinned the 200 is flushed at once, and a failure while
// streaming cannot take it back: the response is aborted instead, which a
// client reads as an unexpected end of the body, and the archive's missing
// `snapshot.json` says the same to whoever extracts what arrived.
func (h *handler) getExport(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	exporter, ok := DatasetFrom(ctx).(substrate.Exporter)
	if !ok {
		writeUnsupported(w, "this service does not export a repository")
		return
	}
	export, err := exporter.Export(ctx)
	if err != nil {
		writeSubstrateError(w, err)
		return
	}
	point := export.Point()
	w.Header().Set("Content-Type", "application/x-tar")
	w.Header().Set("Content-Disposition", mime.FormatMediaType("attachment",
		map[string]string{"filename": fmt.Sprintf("%s-%d.tar", point.Authority, point.Head)}))
	w.Header().Set("Cache-Control", "no-store")
	w.WriteHeader(http.StatusOK)
	// The headers reach the client before the first blob is read, so a
	// client opens its output file against a 200 and a mid-stream failure
	// is a cut body, never a request that got no answer.
	_ = http.NewResponseController(w).Flush()
	if _, err := export.WriteTo(w); err != nil {
		slog.Error("export aborted mid-stream", "repository", point.Authority, "head", point.Head, "error", err)
		panic(http.ErrAbortHandler)
	}
}

File: web/console/src/lib/api/export.ts
/** The recovery export (`GET /api/v1/export`, decision 0069): the
 * repository's directory as of one committed point, as a tar. The token is the
 * whole credential, as it is for every record the archive carries. */

import { envelopeError, rootPath } from "./http"
import { getToken, sessionExpired } from "./session"

const FALLBACK_NAME = "substrate-export.tar"

/** The file name the server offered, or a plain one. */
export function exportFileName(disposition: string | null): string {
  const match = disposition?.match(/filename\*?=(?:UTF-8'')?"?([^";]+)"?/i)
  return match?.[1] ? decodeURIComponent(match[1]) : FALLBACK_NAME
}

/** Read the archive and hand it to the browser as a download. */
export async function downloadExport(): Promise<string> {
  const headers: Record<string, string> = { "X-Substrate-Actor": "console" }
  const token = getToken()
  if (token) headers.Authorization = `Bearer ${token}`
  const res = await fetch(rootPath("export"), { headers })
  if (!res.ok) {
    if (res.status === 401) sessionExpired()
    let body: unknown
    try {
      body = await res.json()
    } catch {
      body = undefined
    }
    throw envelopeError(res.status, body)
  }
  const name = exportFileName(res.headers.get("Content-Disposition"))
  const url = URL.createObjectURL(await res.blob())
  try {
    const link = document.createElement("a")
    link.href = url
    link.download = name
    document.body.appendChild(link)
    link.click()
    link.remove()
  } finally {
    setTimeout(() => URL.revokeObjectURL(url), 60_000)
  }
  return name
}
